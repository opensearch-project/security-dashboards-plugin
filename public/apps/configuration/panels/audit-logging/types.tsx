/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export interface GeneralSettings {
  // layer
  enable_rest?: boolean;
  enable_transport?: boolean;
  disabled_rest_categories?: string[];
  disabled_transport_categories?: string[];

  // attribute
  resolve_bulk_requests?: boolean;
  resolve_indices?: boolean;
  log_request_body?: boolean;
  exclude_sensitive_headers?: boolean;

  // ignore
  ignore_users?: string[];
  ignore_requests?: string[];
}

export interface ComplianceSettings {
  enabled?: boolean;

  // read
  read_metadata_only?: boolean;
  read_ignore_users?: string[];
  read_watched_fields?: {
    [key: string]: string;
  };

  // write
  write_metadata_only?: boolean;
  write_log_diffs?: boolean;
  write_ignore_users?: string[];
  write_watched_indices?: string[];
}

export interface AuditLoggingSettings {
  enabled?: boolean;
  audit?: GeneralSettings;
  compliance?: ComplianceSettings;
}
