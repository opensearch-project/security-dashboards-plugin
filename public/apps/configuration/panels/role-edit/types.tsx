/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export type ComboBoxOptions = EuiComboBoxOptionOption[];
export type FieldLevelSecurityMethod = 'exclude' | 'include';
export interface RoleIndexPermissionStateClass {
  indexPatterns: ComboBoxOptions;
  docLevelSecurity: string;
  fieldLevelSecurityMethod: FieldLevelSecurityMethod;
  fieldLevelSecurityFields: ComboBoxOptions;
  maskedFields: ComboBoxOptions;
  allowedActions: ComboBoxOptions;
}

export enum TenantPermissionType {
  None = '',
  Read = 'r',
  Write = 'w',
  Full = 'rw',
}

export interface RoleTenantPermissionStateClass {
  tenantPatterns: ComboBoxOptions;
  permissionType: TenantPermissionType;
}
