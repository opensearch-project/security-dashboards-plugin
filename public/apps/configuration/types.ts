/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export type ComboBoxOptions = EuiComboBoxOptionOption[];

export type FieldLevelSecurityMethod = 'exclude' | 'include';

export enum ResourceType {
  roles = 'roles',
  users = 'users',
  permissions = 'permissions',
  tenants = 'tenants',
  tenantsManageTab = 'tenantsManageTab',
  tenantsConfigureTab = 'tenantsConfigureTab',
  auth = 'auth',
  auditLogging = 'auditLogging',
}

export enum Action {
  view = 'view',
  create = 'create',
  edit = 'edit',
  duplicate = 'duplicate',
  manage = 'manage',
  configure = 'configure',
}

export enum SubAction {
  mapuser = 'mapuser',
}

export enum PageId {
  dashboardId = 'dashboards',
  visualizationId = 'visualize',
}

export enum TenantPermissionType {
  None = '',
  Read = 'Read only',
  ReadWrite = 'Read and Write',
}

export interface RouteItem {
  name: string;
  href: string;
}

export interface DataObject<T> {
  [key: string]: T;
}

export interface ObjectsMessage<T> {
  total: number;
  data: DataObject<T>;
}

export interface RoleIndexPermission {
  index_patterns: string[];
  dls: string;
  fls: string[];
  masked_fields: string[];
  allowed_actions: string[];
}

export interface RoleIndexPermissionView extends RoleIndexPermission {
  id: number;
}

export interface RoleTenantPatterns {
  tenant_patterns: string[];
}

export interface RoleTenantPermission extends RoleTenantPatterns {
  allowed_actions: string[];
}

export interface RoleTenantPermissionView extends RoleTenantPatterns {
  permissionType: string;
}

export interface RoleUpdate {
  cluster_permissions: string[];
  index_permissions: RoleIndexPermission[];
  tenant_permissions: RoleTenantPermission[];
}

export interface RoleDetail extends RoleUpdate {
  reserved: boolean;
}

export interface RoleMappingDetail {
  backend_roles: string[];
  hosts: string[];
  users: string[];
}

export interface TenantUpdate {
  description: string;
}

export interface TenantName {
  tenant: string;
}

export interface Tenant extends TenantUpdate, TenantName {
  reserved: boolean;
  tenantValue: string;
}

export interface TenantSelect extends TenantName {
  username: string;
}

export interface RoleTenantPermissionDetail extends RoleTenantPermissionView, Tenant {}

export interface UserAttributes {
  [key: string]: string;
}

export interface InternalUser {
  attributes: UserAttributes;
  backend_roles: string[];
}

export interface InternalUserUpdate extends InternalUser {
  password?: string;
}

export interface ActionGroupUpdate {
  allowed_actions: string[];
}

export interface ActionGroupItem extends ActionGroupUpdate {
  reserved: boolean;
  type: 'cluster' | 'index' | 'all' | 'kibana' | undefined;
}

export interface ExpandedRowMapInterface {
  [key: string]: React.ReactNode;
}

export interface FormRowDeps {
  headerText: string;
  optional?: boolean;
  headerSubText?: string;
  helpLink?: string;
  helpText?: string;
}
