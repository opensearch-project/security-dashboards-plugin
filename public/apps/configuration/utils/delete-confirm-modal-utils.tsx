/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState } from 'react';
import { EuiOverlayMask, EuiConfirmModal } from '@elastic/eui';

/**
 *
 * @param handleDelete: [Type: func] delete function which needs to be execute on click of confirm button.
 * @param entity: e.g. role(s), tenant(s), user(s), mapping etc. This will display in confirmation text before deletion.
 * @param customConfirmationText: If you want other than default confirm message, pass it as customConfirmationText.
 */
export function useDeleteConfirmState(
  handleDelete: () => Promise<void>,
  entity: string,
  customConfirmationText?: React.ReactNode
): [() => void, React.ReactNode] {
  const [isDeleteConfirmModalVisible, setIsDeleteConfirmModalVisible] = useState(false);
  const closeDeleteConfirmModal = () => setIsDeleteConfirmModalVisible(false);
  const showDeleteConfirmModal = () => setIsDeleteConfirmModalVisible(true);

  const handleConfirm = async () => {
    await handleDelete();
    closeDeleteConfirmModal();
  };

  let deleteConfirmModal;
  if (isDeleteConfirmModalVisible) {
    deleteConfirmModal = (
      <EuiOverlayMask>
        <EuiConfirmModal
          title="Confirm Delete"
          onCancel={closeDeleteConfirmModal}
          onConfirm={handleConfirm}
          cancelButtonText="Cancel"
          confirmButtonText="Confirm"
          defaultFocusedButton="confirm"
        >
          {customConfirmationText ? (
            customConfirmationText
          ) : (
            <p>Do you really want to delete selected {entity}?</p>
          )}
        </EuiConfirmModal>
      </EuiOverlayMask>
    );
  }
  return [showDeleteConfirmModal, deleteConfirmModal];
}
