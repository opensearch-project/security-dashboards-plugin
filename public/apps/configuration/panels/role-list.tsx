/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState, useEffect, useContext } from 'react';
import {
  EuiFlexGroup,
  EuiText,
  EuiPageHeader,
  EuiTitle,
  EuiPageContent,
  EuiPageContentHeader,
  EuiPageContentHeaderSection,
  EuiFlexItem,
  EuiButton,
  EuiPageBody,
  EuiInMemoryTable,
  EuiBasicTableColumn,
  EuiButtonEmpty,
  EuiSearchBarProps,
  Query,
  EuiLoadingContent,
} from '@elastic/eui';
import { difference } from 'lodash';
import { AppDependencies } from '../../types';
import {
  transformRoleData,
  requestDeleteRoles,
  RoleListing,
  fetchRole,
  fetchRoleMapping,
  buildSearchFilterOptions,
} from '../utils/role-list-utils';
import { Action } from '../types';
import { ResourceType } from '../../../../common';
import { buildHashUrl } from '../utils/url-builder';
import {
  ExternalLink,
  renderCustomization,
  truncatedListView,
  tableItemsUIProps,
} from '../utils/display-utils';
import { showTableStatusMessage } from '../utils/loading-spinner-utils';
import { useDeleteConfirmState } from '../utils/delete-confirm-modal-utils';
import { useContextMenuState } from '../utils/context-menu';
import { DocLinks } from '../constants';
import { DataSourceContext } from '../app-router';
import { SecurityPluginTopNavMenu } from '../top-nav-menu';
import { AccessErrorComponent } from '../access-error-component';

const columns: Array<EuiBasicTableColumn<RoleListing>> = [
  {
    field: 'roleName',
    name: 'Role',
    render: (text: string) => (
      <a href={buildHashUrl(ResourceType.roles, Action.view, text)}>{text}</a>
    ),
    sortable: true,
  },
  {
    field: 'clusterPermissions',
    name: 'Cluster permissions',
    render: truncatedListView(tableItemsUIProps),
    truncateText: true,
  },
  {
    field: 'indexPermissions',
    name: 'Index',
    render: truncatedListView(tableItemsUIProps),
    truncateText: true,
  },
  {
    field: 'internalUsers',
    name: 'Internal users',
    render: truncatedListView(tableItemsUIProps),
  },
  {
    field: 'backendRoles',
    name: 'Backend roles',
    render: truncatedListView(tableItemsUIProps),
  },
  {
    field: 'tenantPermissions',
    name: 'Tenants',
    render: truncatedListView(tableItemsUIProps),
  },
  {
    field: 'reserved',
    name: 'Customization',
    render: (reserved: boolean) => {
      return renderCustomization(reserved, tableItemsUIProps);
    },
  },
];

export function RoleList(props: AppDependencies) {
  const [roleData, setRoleData] = React.useState<RoleListing[]>([]);
  const [errorFlag, setErrorFlag] = React.useState(false);
  const [selection, setSelection] = React.useState<RoleListing[]>([]);
  const [loading, setLoading] = useState(false);
  const [accessErrorFlag, setAccessErrorFlag] = React.useState(false);
  const { dataSource, setDataSource } = useContext(DataSourceContext)!;

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const rawRoleData = await fetchRole(props.coreStart.http, dataSource.id);
        const rawRoleMappingData = await fetchRoleMapping(props.coreStart.http, dataSource.id);
        const processedData = transformRoleData(rawRoleData, rawRoleMappingData);
        setRoleData(processedData);
        setErrorFlag(false);
        setAccessErrorFlag(false);
      } catch (e) {
        console.log(e);
        // requests with existing credentials but insufficient permissions result in 403, remote data-source requests with non-existing credentials result in 400
        if (e.response && [400, 403].includes(e.response.status)) {
          setAccessErrorFlag(true);
        }
        setErrorFlag(true);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [props.coreStart.http, dataSource]);

  const handleDelete = async () => {
    const rolesToDelete: string[] = selection.map((r) => r.roleName);
    try {
      await requestDeleteRoles(props.coreStart.http, rolesToDelete, dataSource.id);
      // Refresh from server (calling fetchData) does not work here, the server still return the roles
      // that had been just deleted, probably because ES takes some time to sync to all nodes.
      // So here remove the selected roles from local memory directly.
      setRoleData(difference(roleData, selection));
      setSelection([]);
    } catch (e) {
      console.log(e);
    } finally {
      closeActionsMenu();
    }
  };

  const [showDeleteConfirmModal, deleteConfirmModal] = useDeleteConfirmState(
    handleDelete,
    'role(s)'
  );

  const actionsMenuItems = [
    <EuiButtonEmpty
      data-test-subj="edit"
      key="edit"
      onClick={() => {
        window.location.href = buildHashUrl(ResourceType.roles, Action.edit, selection[0].roleName);
      }}
      disabled={selection.length !== 1 || selection[0].reserved}
    >
      Edit
    </EuiButtonEmpty>,
    // TODO: Change duplication to a popup window
    <EuiButtonEmpty
      data-test-subj="duplicate"
      key="duplicate"
      onClick={() => {
        window.location.href = buildHashUrl(
          ResourceType.roles,
          Action.duplicate,
          selection[0].roleName
        );
      }}
      disabled={selection.length !== 1}
    >
      Duplicate
    </EuiButtonEmpty>,
    <EuiButtonEmpty
      key="delete"
      color="danger"
      onClick={showDeleteConfirmModal}
      disabled={selection.length === 0 || selection.some((e) => e.reserved)}
    >
      Delete
    </EuiButtonEmpty>,
  ];

  const [actionsMenu, closeActionsMenu] = useContextMenuState('Actions', {}, actionsMenuItems);

  const [searchOptions, setSearchOptions] = useState<EuiSearchBarProps>({});
  const [query, setQuery] = useState<Query | null>(null);
  useEffect(() => {
    setSearchOptions({
      onChange: (arg) => {
        setQuery(arg.query);
        return true;
      },
      filters: [
        {
          type: 'field_value_selection',
          field: 'clusterPermissions',
          name: 'Cluster permissions',
          multiSelect: 'or',
          options: buildSearchFilterOptions(roleData, 'clusterPermissions'),
        },
        {
          type: 'field_value_selection',
          field: 'indexPermissions',
          name: 'Index',
          multiSelect: 'or',
          options: buildSearchFilterOptions(roleData, 'indexPermissions'),
        },
        {
          type: 'field_value_selection',
          field: 'internalUsers',
          name: 'Internal users',
          multiSelect: 'or',
          options: buildSearchFilterOptions(roleData, 'internalUsers'),
        },
        {
          type: 'field_value_selection',
          field: 'backendRoles',
          name: 'Backend roles',
          multiSelect: 'or',
          options: buildSearchFilterOptions(roleData, 'backendRoles'),
        },
        {
          type: 'field_value_selection',
          field: 'tenantPermissions',
          name: 'Tenants',
          multiSelect: 'or',
          options: buildSearchFilterOptions(roleData, 'tenantPermissions'),
        },
        {
          type: 'field_value_selection',
          field: 'reserved',
          name: 'Customization',
          multiSelect: false,
          options: [
            {
              value: true,
              view: renderCustomization(true, tableItemsUIProps),
            },
            {
              value: false,
              view: renderCustomization(false, tableItemsUIProps),
            },
          ],
        },
      ],
    });
  }, [roleData]);

  return (
    <>
      <SecurityPluginTopNavMenu
        {...props}
        dataSourcePickerReadOnly={false}
        setDataSource={setDataSource}
        selectedDataSource={dataSource}
      />
      <EuiPageHeader>
        <EuiTitle size="l">
          <h1>Roles</h1>
        </EuiTitle>
      </EuiPageHeader>
      {loading ? (
        <EuiLoadingContent />
      ) : accessErrorFlag ? (
        <AccessErrorComponent loading={loading} dataSourceLabel={dataSource && dataSource.label} />
      ) : (
        <EuiPageContent>
          <EuiPageContentHeader id="role-table-container">
            <EuiPageContentHeaderSection>
              <EuiTitle size="s">
                <h3>
                  Roles
                  <span className="panel-header-count">
                    {' '}
                    ({Query.execute(query || '', roleData).length})
                  </span>
                </h3>
              </EuiTitle>
              <EuiText size="xs" color="subdued">
                Roles are the core way of controlling access to your cluster. Roles contain any
                combination of cluster-wide permission, index-specific permissions, document- and
                field-level security, and tenants. Then you map users to these roles so that users
                gain those permissions. <ExternalLink href={DocLinks.UsersAndRolesDoc} />
              </EuiText>
            </EuiPageContentHeaderSection>
            <EuiPageContentHeaderSection>
              <EuiFlexGroup>
                <EuiFlexItem>{actionsMenu}</EuiFlexItem>
                <EuiFlexItem>
                  <EuiButton
                    fill
                    href={buildHashUrl(ResourceType.roles, Action.create)}
                    data-test-subj="create-role"
                  >
                    Create role
                  </EuiButton>
                </EuiFlexItem>
              </EuiFlexGroup>
            </EuiPageContentHeaderSection>
          </EuiPageContentHeader>
          <EuiPageBody>
            <EuiInMemoryTable
              data-test-subj="role-list"
              tableLayout={'auto'}
              loading={roleData === [] && !errorFlag}
              columns={columns}
              items={roleData}
              itemId={'roleName'}
              pagination={true}
              selection={{ onSelectionChange: setSelection }}
              sorting={true}
              search={searchOptions}
              error={errorFlag ? 'Load data failed, please check console log for more detail.' : ''}
              message={showTableStatusMessage(loading, roleData)}
            />
          </EuiPageBody>
          {deleteConfirmModal}
        </EuiPageContent>
      )}
    </>
  );
}
