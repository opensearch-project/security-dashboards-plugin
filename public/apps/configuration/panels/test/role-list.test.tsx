/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { RoleList } from '../role-list';
import { mount, shallow } from 'enzyme';
import React from 'react';
import { EuiInMemoryTable, EuiTableFieldDataColumnType } from '@elastic/eui';
import { buildHashUrl } from '../../utils/url-builder';
import { Action } from '../../types';
import { ResourceType } from '../../../../../common';
import { RoleListing } from '../../utils/role-list-utils';
import { useDeleteConfirmState } from '../../utils/delete-confirm-modal-utils';

jest.mock('../../utils/role-list-utils');
jest.mock('../../utils/display-utils', () => ({
  ...jest.requireActual('../../utils/display-utils'),
  ExternalLink: jest.fn(),
}));
jest.mock('../../utils/context-menu', () => ({
  useContextMenuState: jest
    .fn()
    .mockImplementation((buttonText, buttonProps, children) => [children, jest.fn()]),
}));
jest.mock('../../utils/delete-confirm-modal-utils', () => ({
  useDeleteConfirmState: jest.fn().mockReturnValue([jest.fn(), '']),
}));
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn().mockReturnValue({ dataSource: { id: 'test' }, setDataSource: jest.fn() }), // Mock the useContext hook to return dummy datasource and setdatasource function
}));

// eslint-disable-next-line
const mockRoleListUtils = require('../../utils/role-list-utils');

describe('Role list', () => {
  const setState = jest.fn();
  const mockCoreStart = {
    http: 1,
  };

  beforeEach(() => {
    jest.spyOn(React, 'useState').mockImplementation((initialValue) => [initialValue, setState]);
  });

  it('Render empty', () => {
    const mockRoleListingData = [
      {
        roleName: 'role_1',
        reserved: true,
        clusterPermission: ['readonly'],
        indexPermission: ['data1'],
        tenantPermission: ['tenant1'],
        internaluser: [],
        backendRoles: [],
      },
    ];

    mockRoleListUtils.transformRoleData = jest.fn().mockReturnValue(mockRoleListingData);

    const component = shallow(
      <RoleList
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );

    expect(component.find(EuiInMemoryTable).prop('items').length).toBe(0);
  });

  it('Render error', (done) => {
    jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
    // Hide the error message
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    mockRoleListUtils.fetchRole.mockImplementationOnce(() => {
      throw Error();
    });

    shallow(
      <RoleList
        coreStart={{} as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );

    process.nextTick(() => {
      // Expect setting error to true
      expect(setState).toHaveBeenCalledWith(true);
      done();
    });
  });

  it('Render data', (done) => {
    jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());

    const mockRoleListingData = [
      {
        roleName: 'role_1',
        reserved: true,
        clusterPermission: ['readonly'],
        indexPermission: ['data1'],
        tenantPermission: ['tenant1'],
        internaluser: [],
        backendRoles: [],
      },
    ];

    mockRoleListUtils.transformRoleData = jest.fn().mockReturnValue(mockRoleListingData);

    shallow(
      <RoleList
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );

    process.nextTick(() => {
      expect(mockRoleListUtils.fetchRole).toHaveBeenCalledTimes(1);
      expect(mockRoleListUtils.fetchRoleMapping).toHaveBeenCalledTimes(1);
      expect(mockRoleListUtils.transformRoleData).toHaveBeenCalledTimes(1);
      expect(setState).toHaveBeenCalledWith(mockRoleListingData);
      done();
    });
  });

  it('delete role', (done) => {
    shallow(
      <RoleList
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    const deleteFunc = useDeleteConfirmState.mock.calls[0][0];

    deleteFunc();

    process.nextTick(() => {
      expect(mockRoleListUtils.requestDeleteRoles).toBeCalled();
      done();
    });
  });

  it('error occurred while deleting the role', () => {
    (mockRoleListUtils.requestDeleteRoles as jest.Mock).mockImplementationOnce(() => {
      throw new Error();
    });
    const spy = jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    shallow(
      <RoleList
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    const deleteFunc = useDeleteConfirmState.mock.calls[0][0];

    deleteFunc();
    expect(spy).toBeCalled();
  });

  describe('Action menu click', () => {
    let component;
    const mockRoleListingData: RoleListing = {
      roleName: 'role_1',
      reserved: true,
      clusterPermissions: ['readonly'],
      indexPermissions: ['data1'],
      tenantPermissions: ['tenant1'],
      internalUsers: [],
      backendRoles: [],
    };
    beforeEach(() => {
      jest.spyOn(React, 'useState').mockRestore();
      jest
        .spyOn(React, 'useState')
        .mockImplementationOnce(() => [[mockRoleListingData], jest.fn()])
        .mockImplementationOnce(() => [false, jest.fn()])
        .mockImplementationOnce(() => [[mockRoleListingData], jest.fn()])
        .mockImplementationOnce(() => [false, jest.fn()]);
      component = shallow(
        <RoleList
          coreStart={mockCoreStart as any}
          navigation={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
    });

    it('Edit click', () => {
      component.find('[data-test-subj="edit"]').simulate('click');
      expect(window.location.hash).toBe(
        buildHashUrl(ResourceType.roles, Action.edit, mockRoleListingData.roleName)
      );
    });

    it('Duplicate click', () => {
      component.find('[data-test-subj="duplicate"]').simulate('click');
      expect(window.location.hash).toBe(
        buildHashUrl(ResourceType.roles, Action.duplicate, mockRoleListingData.roleName)
      );
    });
  });

  describe('Render columns', () => {
    beforeEach(() => {
      jest.spyOn(React, 'useState').mockRestore();
      jest
        .spyOn(React, 'useState')
        .mockImplementationOnce(() => [[], jest.fn()])
        .mockImplementationOnce(() => [false, jest.fn()])
        .mockImplementationOnce(() => [[], jest.fn()])
        .mockImplementationOnce(() => [false, jest.fn()]);
    });
    it('render role name column', () => {
      const wrapper = shallow(
        <RoleList
          coreStart={mockCoreStart as any}
          navigation={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
      const roleList = wrapper.find('[data-test-subj="role-list"]');
      const component = mount(<>{roleList}</>);
      const columns = component.prop<Array<EuiTableFieldDataColumnType<RoleListing>>>('columns');

      const roleNameRenderer = columns[0].render as (roleName: string) => JSX.Element;
      const Container = (props: { roleName: string }) => roleNameRenderer(props.roleName);
      const result = shallow(<Container roleName={'role1'} />);
      expect(result).toMatchSnapshot();
    });

    it('render Customization column', () => {
      const wrapper = shallow(
        <RoleList
          coreStart={mockCoreStart as any}
          navigation={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
      const roleList = wrapper.find('[data-test-subj="role-list"]');
      const component = mount(<>{roleList}</>);
      const columns = component.prop<Array<EuiTableFieldDataColumnType<RoleListing>>>('columns');

      const customizationRenderer = columns[6].render as (reserved: boolean) => JSX.Element;
      const Container = (props: { reserved: boolean }) => customizationRenderer(props.reserved);
      const result = shallow(<Container reserved={true} />);
      expect(result).toMatchSnapshot();
    });
  });

  describe('AccessError component', () => {
    let component;
    beforeEach(() => {
      jest.spyOn(React, 'useState').mockRestore();
      jest
        .spyOn(React, 'useState')
        .mockImplementationOnce(() => [[], jest.fn()])
        .mockImplementationOnce(() => [false, jest.fn()])
        .mockImplementationOnce(() => [[], jest.fn()])
        .mockImplementationOnce(() => [false, jest.fn()])
        .mockImplementationOnce(() => [true, jest.fn()]);
      component = shallow(
        <RoleList
          coreStart={mockCoreStart as any}
          navigation={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
    });

    it('should load access error component', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
