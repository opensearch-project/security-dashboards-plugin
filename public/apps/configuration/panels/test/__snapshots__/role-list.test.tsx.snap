// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Role list AccessError component should load access error component 1`] = `
<Fragment>
  <Memo()
    config={Object {}}
    coreStart={
      Object {
        "http": 1,
      }
    }
    dataSourcePickerReadOnly={false}
    navigation={Object {}}
    params={Object {}}
    selectedDataSource={
      Object {
        "id": "test",
      }
    }
    setDataSource={[MockFunction]}
  />
  <EuiPageHeader>
    <EuiTitle
      size="l"
    >
      <h1>
        Roles
      </h1>
    </EuiTitle>
  </EuiPageHeader>
  <EuiPageContent>
    <EuiPageContentHeader
      id="role-table-container"
    >
      <EuiPageContentHeaderSection>
        <EuiTitle
          size="s"
        >
          <h3>
            Roles
            <span
              className="panel-header-count"
            >
               
              (
              0
              )
            </span>
          </h3>
        </EuiTitle>
        <EuiText
          color="subdued"
          size="xs"
        >
          Roles are the core way of controlling access to your cluster. Roles contain any combination of cluster-wide permission, index-specific permissions, document- and field-level security, and tenants. Then you map users to these roles so that users gain those permissions. 
          <mockConstructor
            href="https://opensearch.org/docs/latest/security-plugin/access-control/users-roles/"
          />
        </EuiText>
      </EuiPageContentHeaderSection>
      <EuiPageContentHeaderSection>
        <EuiFlexGroup>
          <EuiFlexItem>
            <EuiButtonEmpty
              data-test-subj="edit"
              disabled={true}
              key="edit"
              onClick={[Function]}
            >
              Edit
            </EuiButtonEmpty>
            <EuiButtonEmpty
              data-test-subj="duplicate"
              disabled={true}
              key="duplicate"
              onClick={[Function]}
            >
              Duplicate
            </EuiButtonEmpty>
            <EuiButtonEmpty
              color="danger"
              disabled={true}
              key="delete"
              onClick={[MockFunction]}
            >
              Delete
            </EuiButtonEmpty>
          </EuiFlexItem>
          <EuiFlexItem>
            <EuiButton
              data-test-subj="create-role"
              fill={true}
              href="#/roles/create"
            >
              Create role
            </EuiButton>
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiPageContentHeaderSection>
    </EuiPageContentHeader>
    <EuiPageBody>
      <EuiInMemoryTable
        columns={
          Array [
            Object {
              "field": "roleName",
              "name": "Role",
              "render": [Function],
              "sortable": true,
            },
            Object {
              "field": "clusterPermissions",
              "name": "Cluster permissions",
              "render": [Function],
              "truncateText": true,
            },
            Object {
              "field": "indexPermissions",
              "name": "Index",
              "render": [Function],
              "truncateText": true,
            },
            Object {
              "field": "internalUsers",
              "name": "Internal users",
              "render": [Function],
            },
            Object {
              "field": "backendRoles",
              "name": "Backend roles",
              "render": [Function],
            },
            Object {
              "field": "tenantPermissions",
              "name": "Tenants",
              "render": [Function],
            },
            Object {
              "field": "reserved",
              "name": "Customization",
              "render": [Function],
            },
          ]
        }
        data-test-subj="role-list"
        error=""
        itemId="roleName"
        items={Array []}
        loading={false}
        message="No items found"
        pagination={true}
        responsive={true}
        search={Object {}}
        selection={
          Object {
            "onSelectionChange": [MockFunction],
          }
        }
        sorting={true}
        tableLayout="auto"
      />
    </EuiPageBody>
  </EuiPageContent>
</Fragment>
`;

exports[`Role list Render columns render Customization column 1`] = `
<EuiFlexGroup
  alignItems="center"
  gutterSize="xs"
>
  <EuiFlexItem
    grow={false}
  >
    <EuiIcon
      type="lock"
    />
  </EuiFlexItem>
  <EuiFlexItem
    grow={false}
  >
    <EuiText
      className="table-items"
      data-test-subj="filter-reserved"
    >
      Reserved
    </EuiText>
  </EuiFlexItem>
</EuiFlexGroup>
`;

exports[`Role list Render columns render role name column 1`] = `
<a
  href="#/roles/view/role1"
>
  role1
</a>
`;
